ledger/fullblock_perf_test.go

```diff
@@ -0,0 +1,331 @@
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/algorand/go-algorand/agreement"
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/data/transactions/verify"
+	"github.com/algorand/go-algorand/ledger/internal"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+func setupEnv(dbTempDir, name string, numAccts int) (l0,
+	l1 *Ledger, creator basics.Address, accts []basics.Address, aIdxs []uint64, err error) {
+	dbName := fmt.Sprintf("%s.%d", name, crypto.RandUint64())
+	dbPrefix := filepath.Join(dbTempDir, dbName)
+
+	genesisInitState := getInitState()
+
+	// Use future protocol
+	genesisInitState.Block.BlockHeader.GenesisHash = crypto.Digest{}
+	genesisInitState.Block.CurrentProtocol = protocol.ConsensusFuture
+	genesisInitState.GenesisHash = crypto.Digest{1}
+	genesisInitState.Block.BlockHeader.GenesisHash = crypto.Digest{1}
+
+	creator = basics.Address{}
+	_, err = rand.Read(creator[:])
+	if err != nil {
+		return nil, nil, basics.Address{}, nil, nil, err
+	}
+	genesisInitState.Accounts[creator] = basics.MakeAccountData(basics.Offline, basics.MicroAlgos{Raw: 12345678900000})
+	for i := 0; i < numAccts; i++ {
+		acct := basics.Address{}
+		_, err = rand.Read(acct[:])
+		if err != nil {
+			return nil, nil, basics.Address{}, nil, nil, err
+		}
+		genesisInitState.Accounts[acct] = basics.MakeAccountData(basics.Offline, basics.MicroAlgos{Raw: 1234567890})
+		accts = append(accts, acct)
+	}
+
+	// open ledger
+	const inMem = false
+	cfg := config.GetDefaultLocal()
+	cfg.Archival = true
+	l0, err = OpenLedger(logging.Base(), dbPrefix, inMem, genesisInitState, cfg)
+	if err != nil {
+		return nil, nil, basics.Address{}, nil, nil, err
+	}
+
+	// open second ledger
+	dbName = fmt.Sprintf("%s.%d.2", name, crypto.RandUint64())
+	dbPrefix = filepath.Join(dbTempDir, dbName)
+	l1, err = OpenLedger(logging.Base(), dbPrefix, inMem, genesisInitState, cfg)
+	if err != nil {
+		return nil, nil, basics.Address{}, nil, nil, err
+	}
+
+	blk := genesisInitState.Block
+	blk.BlockHeader.Round++
+	blk.BlockHeader.TimeStamp += int64(crypto.RandUint64() % 100 * 1000)
+	blk.BlockHeader.GenesisID = "x"
+	cert := agreement.Certificate{}
+
+	err = l0.AddBlock(blk, cert)
+	if err != nil {
+		return nil, nil, basics.Address{}, nil, nil, err
+	}
+	err = l1.AddBlock(blk, cert)
+	if err != nil {
+		return nil, nil, basics.Address{}, nil, nil, err
+	}
+
+	newBlk := bookkeeping.MakeBlock(blk.BlockHeader)
+	eval, err := l0.StartEvaluator(newBlk.BlockHeader, 5000, 0)
+	if err != nil {
+		return nil, nil, basics.Address{}, nil, nil, err
+	}
+
+	for i, acct := range accts {
+		tx := createAssetTransaction(uint64(i), 1, creator, crypto.Digest{1})
```

Comment:
maybe we make asset creation as part of regular block validate mix, and not pre-init stage?